onnxruntime/test/mlas/unittest.cpp

```diff
@@ -2495,6 +2495,206 @@ class MlasComputeExpTest : public MlasTestBase
+
+        for (size_t n = 0; n < N; n++) {
+            int diff = (int)OutputC[n] - (int)OutputReference[n];
+            if (diff < -1 || diff > 1) {
+                printf("TestVectorScalar exp difference @ %u of %u, %d + %d => %d (expecting %d)\n",
+                        static_cast<unsigned>(n), static_cast<unsigned>(N),
+                        static_cast<int>(InputA[n]), static_cast<int>(InputB[0]),
+                        static_cast<int>(OutputC[n]), static_cast<int>(OutputReference[n]));
+            }
+        }
+
+        BufferInputA.ReleaseBuffer();
+        BufferInputB.ReleaseBuffer();
+        BufferOutput.ReleaseBuffer();
+        BufferOutputReference.ReleaseBuffer();
+    }
+
+    void
+    TestScalarVector(
+        size_t N,
+        float ScaleA,
+        int32_t ZeroPointA,
+        float ScaleB,
+        int32_t ZeroPointB,
+        float ScaleC,
+        int32_t ZeroPointC
+        )
+    {
+        T* InputA = BufferInputA.GetBuffer(1);
+        T* InputB = BufferInputB.GetBuffer(N);
+        T* OutputC = BufferOutput.GetBuffer(N);
+        T* OutputReference = BufferOutputReference.GetBuffer(N);
+
+        constexpr int MinimumValue = (int)std::numeric_limits<T>::min();
+        constexpr int MaximumValue = (int)std::numeric_limits<T>::max();
+        std::default_random_engine generator(static_cast<unsigned>(N));
+        std::uniform_int_distribution<int> distribution(MinimumValue, MaximumValue);
+
+        InputA[0] = static_cast<T>(distribution(generator));
+        for (size_t n = 0; n < N; n++) {
+            InputB[n] = static_cast<T>(distribution(generator));
+            OutputReference[n] = QLinearAddScalar(InputA[0], ScaleA, ZeroPointA, InputB[n], ScaleB, ZeroPointB, ScaleC, ZeroPointC);
+        }
+
+        MlasQLinearAdd(InputA, ScaleA, ZeroPointA, InputB, ScaleB, ZeroPointB, ScaleC, ZeroPointC, OutputC, 1, N);
+
+        for (size_t n = 0; n < N; n++) {
+            int diff = (int)OutputC[n] - (int)OutputReference[n];
+            if (diff < -1 || diff > 1) {
+                printf("TestScalarVector exp difference @ %u of %u, %d + %d => %d (expecting %d)\n",
+                        static_cast<unsigned>(n), static_cast<unsigned>(N),
+                        static_cast<int>(InputA[0]), static_cast<int>(InputB[n]),
+                        static_cast<int>(OutputC[n]), static_cast<int>(OutputReference[n]));
+            }
+        }
+
+        BufferInputA.ReleaseBuffer();
+        BufferInputB.ReleaseBuffer();
+        BufferOutput.ReleaseBuffer();
+        BufferOutputReference.ReleaseBuffer();
+    }
+
+public:
+    void
+    ExecuteShort(
+        void
+        ) override;
+};
+
+template<>
+void
+MlasQLinearAddExpTest<uint8_t>::ExecuteShort(
+    void
+    )
+{
+    for (size_t n = 1; n < 128; n++) {
+        TestVectorVector(n, 10.f, 128, 10.f, 128, 20.f, 128);
+        TestVectorScalar(n, 10.f, 128, 10.f, 128, 20.f, 128);
```

Comment:
Why did we end up with needing a VectorScalar and ScalarVector form? Couldn't the QLinearAdd op have just swapped scalar + vector to be vector + scalar and save some more code? Maybe something for 1.5 if that's too much for now.

Reply:
done